pkg/controller/projectreference/projectreference_adapter.go

```diff
@@ -0,0 +1,341 @@
+	}
+	return &ReferenceAdapter{
+		projectClaim:     projectClaim,
+		projectReference: projectReference,
+		logger:           logger,
+		kubeClient:       client,
+		gcpClient:        gcpClient,
+	}, nil
+}
+
+func getMatchingClaimLink(projectReference *gcpv1alpha1.ProjectReference, client client.Client) (*gcpv1alpha1.ProjectClaim, error) {
+	projectClaim := &gcpv1alpha1.ProjectClaim{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: projectReference.Spec.ProjectClaimCRLink.Name, Namespace: projectReference.Spec.ProjectClaimCRLink.Namespace}, projectClaim)
+	if err != nil {
+		return &gcpv1alpha1.ProjectClaim{}, err
+
+	}
+	return projectClaim, nil
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapter) updateProjectID() error {
+	guid := uuid.New().String()
+	hashing := sha1.New()
+	hashing.Write([]byte(guid))
+	uuidsum := fmt.Sprintf("%x", hashing.Sum(nil))
+	shortuuid := uuidsum[0:26]
+	r.projectReference.Spec.GCPProjectID = "osd-" + shortuuid
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// updateProjectID updates the ProjectReference with a unique ID for the ProjectID
+func (r *ReferenceAdapter) clearProjectID() error {
+	r.projectReference.Spec.GCPProjectID = ""
+	return r.kubeClient.Update(context.TODO(), r.projectReference)
+}
+
+// checkRequirements checks that region is supported
+func (r *ReferenceAdapter) checkRequirements() error {
+	if _, ok := supportedRegions[r.projectClaim.Spec.Region]; !ok {
+		return operrors.ErrRegionNotSupported
+	}
+	return nil
+}
+
+func (r *ReferenceAdapter) createProject(parentFolderID string) error {
+	// Get existing projects
+	projects, err := r.gcpClient.ListProjects()
+	if err != nil {
+		return err
+	}
+
+	projectMap := convertProjectsToMap(projects)
+
+	if project, ok := projectMap[r.projectReference.Spec.GCPProjectID]; ok {
+		switch project.LifecycleState {
+		case "ACTIVE":
+			r.logger.Info("Project lifecycleState == ACTIVE")
+			return nil
+		case "DELETE_REQUESTED":
+			return operrors.ErrInactiveProject
+		default:
+			r.logger.Error(operrors.ErrUnexpectedLifecycleState, "Unexpected LifecycleState", project.LifecycleState)
+			return operrors.ErrUnexpectedLifecycleState
+
+		}
+	}
+
+	r.logger.Info("Creating Project")
+	// If we cannot create the project clear the projectID from spec so we can try again with another unique key
+	_, err = r.gcpClient.CreateProject(parentFolderID)
+	if err != nil {
+		r.logger.Error(err, "could not create project", "Parent Folder ID", parentFolderID, "Requested Project ID", r.projectReference.Spec.GCPProjectID)
+		r.logger.Info("Clearing gcpProjectID from ProjectReferenceSpec")
+		err = r.clearProjectID()
+		if err != nil {
+			return err
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReferenceAdapter) configureAPIS() error {
+	billingAccount, err := util.GetBillingAccountFromSecret(r.kubeClient, operatorNamespace, orgGcpSecretName)
```

Comment:
This needs to be taken from configmap before merge, per current HEAD.

Reply:
This is a good point. For the time being since I have already tested and created testing for the codebase as is. I added the this function back to the util package. We should make a separate card to remove these from the ProjectReference controller. Since there will be a few items refactored around the new configmap.

Reply:
I think the billing account doesn't exist anymore in the secret since we switched to the service accounts + config map

Reply:
In that case I need to recreate the secret on int with billing account re–added, before merge. I'll add it to my todo.

Reply:
See https://github.com/rafael-azevedo/gcp-project-operator-1/pull/1